include/kokkos-utils/callbacks/TimerListener.hpp

```diff
@@ -0,0 +1,209 @@
+ *
+ * The start event is selected by requiring it to be of type @p BeginEventType and
+ * to be a match for the passed matcher. The stop event must be of type @p EndEventType
+ * and have the same ID as the begin event.
+ *
+ * The underlying timer @ref timer is a @ref Kokkos::utils::timer::Timer, instantiated
+ * for the execution space @p Exec. This class stores the execution space
+ * instance passed to the constructor, and uses it as the one passed to @ref timer.
+ *
+ * This class is well-suited for timing the execution of a parallel workload, such as a
+ * @c Kokkos parallel for, parallel reduce, parallel scan, or deep copy operation, to which
+ * an execution space instance is passed. In such a situation, the execution space instance
+ * passed to this class's constructor should be the same as the one passed to the workload.
+ */
+template <
+    Kokkos::utils::callbacks::BeginEvent BeginEventType, Kokkos::utils::callbacks::EndEvent EndEventType,
+    Kokkos::utils::callbacks::Matcher<Kokkos::Impl::type_list<BeginEventType>> MatcherType,
+    Kokkos::utils::concepts::ExecutionSpace Exec
+>
+class TimerListener<BeginEventType, EndEventType, MatcherType, Exec>
+{
+public:
+    static constexpr uint64_t invalid_event_id = Kokkos::Experimental::finite_max_v<uint64_t>;
+
+public:
+    template <typename T> requires (std::same_as<std::remove_cvref_t<T>, MatcherType>)
+    TimerListener(Exec exec_, T&& matcher_)
+      : exec(std::move(exec_)),
+        dev_id(Kokkos::Tools::Experimental::device_id(exec)),
+        matcher(std::forward<T>(matcher_)) {}
+
+    /**
+     * @todo Consider optimizing the implementation to avoid repeating the regex many
+     *       times after the timer is already closed.
+     *
+     * @todo Allow the possibility to combine multiple matchers with logical operators, so that
+     *       we can match e.g. a begin event with a specific name within a specific section.
+     */
+    void operator()(const BeginEventType& event) {
+        if (event.dev_id != this->dev_id) return;
+
+        if (matcher(event)) {
+            if (this->connected() || this->closed()) Kokkos::abort("Listener cannot match a begin event twice.");
+
+            this->timer.reset(this->exec);
+            this->event_id = event.event_id;
+        }
+    }
+
+    void operator()(const EndEventType& event) {
+        if (this->event_id == event.event_id)
+        {
+            this->timer.stop(this->exec);
+            this->is_closed = true;
+        }
+    }
+
+    //! When the listener has matched the begin, but not yet the end event, it is said to be "connected".
+    bool connected() const { return (this->event_id != invalid_event_id) && !this->closed(); }
+
+    //! When the listener has matched both the begin and end events, it is said to be "closed".
+    bool closed() const { return this->is_closed; }
+
+    //! Open the listener so that it can be reused to match a begin event again.
+    void open()
```

Comment:
This is new. Not sure to see:
- why we would reuse a matcher
- why you call this `open` and not `reset` ?

Reply:
It's new indeed! 

I added it to make things work when I used it in a benchmark with more than 1 iteration. The `SetUp` method of the benchmark creates the `TimerListener`. And then the benchmark loops over the `run` method, and so it iterates  through the begin/end, begin/end, ....

I used `open` instead of `reset` because we already use `reset` in the underlying `Timer`. And there, `reset` is not really "resetting" things, but rather recording the start event. Not sure what to do best. We could keep `open` here, which goes well with the `closed`. Another option would be to rename `reset` as `start` in `Timer` and then rename `open` here as `reset`.

Reply:
Agreed with the last suggestion:
- rename `Timer::reset` to `Timer::start`
- rename `open` to `reset`

Reply:
Done too!